### PR TITLE
KV support

### PR DESCRIPTION
**Pull request by Baekalfen**

### CODE REVIEW

**Review by safl — CHANGES_REQUESTED** [review]
Thanks for this!

In addition to adressing:

- [ ] GitHUB Actions breaking
- [ ] Tests utilizing kvs-configured device
- [ ] Enable tests

Then I have added a couple of questions inline along with a couple of minor things that would be good to change.

**Comment by Baekalfen** [comment]
> Thanks for this!
> 
> In addition to addressing:
> 
> * [ ]  Rebase and resolve git-conflicts
> * [ ]  GitHUB Actions breakage introduced by the changes on this PR
> * [ ]  Ensure that tests are utilizing the recently added KV-namespace
>   
>   * see: [Add KV and FDP enabled namespaces #268](https://github.com/OpenMPDK/xNVMe/pull/268)
> * [ ]  Enable kv-tests that were previously disabled due to "not implemented"
>   
>   * see: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_kvs.py
>   * remove: pytest.skip(allow_module_level=True, reason="Not implemented")
> 
> Then I have added a couple of questions inline along with a couple of minor things that would be good to change.

I'll take a look at this.

**Comment by safl** [comment]
Yesterday the self-hosted xNVMe github-actions-runners were "migrated" from one hosting provider to another.
Consequently, then the changes described on https://github.com/OpenMPDK/xNVMe/pull/276 were needed.

To avoid blocking progress on this PR (https://github.com/OpenMPDK/xNVMe/pull/270), due to CI-failures, then this PR was rebased.

Curiously, the rebase of this PR shows in a surprising way on the history. That is, the two commits introduced on https://github.com/OpenMPDK/xNVMe/pull/276 looks like they are part of the PR, as the first two commits. Even though they have the same SHAs as the commits on ``next``. It seems inconsequential, e.g. something cosmetically wrong in the GitHUB UI, regardless, writing this to let you know.

File: include/libxnvme_kvs.h [hunk 1]
```diff
@@ -39,7 +40,7 @@ extern "C" {
  */
 int
 xnvme_kvs_retrieve(struct xnvme_cmd_ctx *ctx, uint32_t nsid, const void *key, uint8_t key_len,
-		   uint8_t opt, const void *dbuf, size_t dbuf_nbytes);
+		   const void *dbuf, uint32_t dbuf_nbytes, uint8_t opt);
```
[the remark this answers is not in the document]
Comment (safl) [reply include/libxnvme_kvs.h]: > 1. The original intend was to 'fixup' (i.e. squash without commit message) the commit on the first KV commit.
> 

I see. Since the PR is a continuation of Eduard's work, then I would prefer that his original work is kept as-is and then, that the changes to that work are described by commits "on top".  There is a bit of history-fiddling to achieve that, e.g. either split the commit currently named "Squash this with first commit " into individual commits, for the changes in that commit which looks like it at least does:

* Change license from APACHE-2 to BSD-3-Clause
* Wrap definitions in enum
* Change signature of functions "x, y, z" dbuf -> vuf
* Change signature: move opt to end
* Re-factor ``kvs_put_key_into_cmd`` to ``kvs_cmd_set_key``

The most helpful part of making such as split is that one can read the motivations behind the changes.

In case  you do not want to spend time making this split, then just let me know, I will then do it and force-push the changes to your branch, for doing so, let me know when I can push without stomping on your branch.

>     2. This can be changed
> 

Great, thanks!

>     3. It was either because we wanted to keep the order of the other arguments the same across different API-function signatures, or we found an example that also had `opts` at the end.

I see. Thanks for clarifying. It would be great if the motivation was part of the commit as noted above.

File: python/xnvme-cy-bindings/xnvme/cython_bindings/tests/test_key_value.py
[the remark this answers is not in the document]
Comment (safl) [reply python/xnvme-cy-bindings/xnvme/cython_bindings/tests/test_key_value.py:1]: > 1. The real pytests came first, and do a more thorough testing. But you didn't like that some of the features were exclusively tested in pytests, so I added the other code too.
> 

From my understanding then the tests in ``toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_kvs.py`` exercise the C API via the ``kvs`` cli-tool and thus covers testing of the ``C API`` as well as the ``kvs`` command-line tool. Whereas the tests in ``python/xnvme-cy-bindings/xnvme/cython_bindings/tests/test_key_value.py`` **only** exercise the C API.

To me that is not as you describe it "more thorough testing", rather less so. So, please help me understand how the tests in ``python/xnvme-cy-bindings/xnvme/cython_bindings/tests/test_key_value.py`` are more thorough. Explicitly, what do they cover which ``toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_kvs.py`` do not? 

>     2. It would require some work to polish off the auto-generated code for ctypes.

Do you know, explicitly, what is missing from the ctypes bindings? That is, what does "polishing" the ctypes-bindings look like in terms of adding to or changing them?

File: include/libxnvme_kvs.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (Baekalfen) [reply include/libxnvme_kvs.h]: You're very welcome to do the split. Then you'll also get it how you want it.

File: python/xnvme-cy-bindings/xnvme/cython_bindings/tests/test_key_value.py
[the remark this answers is not in the document]
Comment (Baekalfen) [reply python/xnvme-cy-bindings/xnvme/cython_bindings/tests/test_key_value.py:1]: The tests in `test_key_value.py` check an initial condition is met, modifies the state and verifies the change to conclude the test. They also verify boundary conditions like maximum key size, multiple/maximum values returned etc. and how those cases are handled.

The tests in `test_kvs.py` only rudimentarily check that the command-line tool didn't return an error code, but doesn't perform any validation of the designated action. It doesn't for example validate/verify the retrieved value of any key.

The tests in `tests/kvs.c` does verify the retrieved value, but only in one instance. No negative tests (checking error-paths and boundaries), no tests of multiple returned values (list command) and only one key/value tested (we need to see we can get different values for different keys).


As for the ctypes, you'll have to do something to support functions returning structs (not struct pointers) as these are ignored by the generator. As I haven't worked with them, I don't know how well the rest works.

File: include/libxnvme_kvs.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (safl) [reply include/libxnvme_kvs.h]: Ok, done. Without alterations to changes in the commits, just a history-rewrite splitting the commit ``Squash this with first commit`` into the following commits:

* xnvmec/kvs: add missing ','
* tests/kvs: adjust to changes in xnvmec-options
* api/xnvmec: add XNVMEC_OPT_KV_STORE_COMPRESS and adjust format
* kvs: rename helper from kvs_put_key_into_cmd() to kvs_cmd_set_key()
* api/kvs,tests/kvs,tools/kvs: change license to BSD-3-Clause
* api/kvs: move KVS-related definitions from spec to kvs
* api/kvs: change signature of xnvme_kvs_{retrieve,store,list}

File: python/xnvme-cy-bindings/xnvme/cython_bindings/tests/test_key_value.py
[the remark this answers is not in the document]
Comment (safl) [reply python/xnvme-cy-bindings/xnvme/cython_bindings/tests/test_key_value.py:1]: > 

Thanks for elaborating, this is helpful.

**Comment by safl** [comment]
Hi @Baekalfen , yet-another-fix for CI is now on ``next``. Looks like this PR has not been affected, yet, it is required now.
There is a conflict as the tests have moved from ``toolbox/cijoe-pkg-xnvme/.../tests/...``` to ``cijoe/tests/...```.
Let me know if you want be to resolve it and force push an update.

File: lib/xnvmec.c [hunk 2]
```diff
@@ -969,6 +969,11 @@ static struct xnvmec_opt_attr xnvmec_opts[] = {
 		.descr = "If set, hint to use range as an integral unit when reading.",
 	},
 
+	{.opt = XNVMEC_OPT_KV_STORE_COMPRESS,
+	 .vtype = XNVMEC_OPT_VTYPE_NUM,
+	 .name = "compress",
+	 .descr = "KV Store Option to compress value on device"},
+
```
Comment (karlowich) [inline lib/xnvmec.c:976]: The format of this is wrong, and it should be moved down under the rest of the KV options, so it resembles the order from libxnvmec.h:
```
const char *kv_key;
const char *kv_val;
bool kv_store_add;
bool kv_store_update;
bool kv_store_compress;
```

File: python/xnvme-cy-bindings/xnvme/cython_bindings/tests/conftest.py [hunk 3]
```diff
@@ -51,15 +51,28 @@ def dev(opts, request):
     opts.rdwr = 1
 
     device = xnvme.xnvme_dev_open(DEVICE_PATH, opts)
+    # if not conventional_namespace:
+    #     pytest.skip('Device must use conventional namespace')
```
Comment (karlowich) [inline python/xnvme-cy-bindings/xnvme/cython_bindings/tests/conftest.py:55]: Is there any reason to keep this commented code?

File: python/xnvme-cy-bindings/xnvme/cython_bindings/tests/conftest.py [hunk 4]
```diff
@@ -51,15 +51,28 @@ def dev(opts, request):
...
     yield device
     device = xnvme.xnvme_dev_close(device)
 
 
+@pytest.fixture
+def dev_kv(dev):
+    # if not key_value_namespace:
+    #     pytest.skip('Device must support key-value namespace')
```
Comment (karlowich) [inline python/xnvme-cy-bindings/xnvme/cython_bindings/tests/conftest.py:63]: Is there any reason to keep this commented code?

**Review by karlowich — COMMENTED** [review]
I have provided a bunch of nitpicks.
Should I make a separate commit implementing them? 
@Baekalfen @safl What do you think?

File: lib/xnvmec.c (hunk 2 above)
Comment (Baekalfen) [reply lib/xnvmec.c:976]: It actually seems like it's duplicate of line 1015. Probably from bad rebase.

File: python/xnvme-cy-bindings/xnvme/cython_bindings/tests/conftest.py (hunk 3 above)
Comment (Baekalfen) [reply python/xnvme-cy-bindings/xnvme/cython_bindings/tests/conftest.py:55]: Not exactly, it served as a todo, but it's afaik not something that can just be fixed.

File: include/libxnvme_spec.h [hunk 5]
```diff
@@ -92,6 +92,16 @@ enum xnvme_spec_status_code_type {
 	XNVME_STATUS_CODE_TYPE_VENDOR  = 0x7, ///< Vendor specific
 };
 
+/**
+ * NVMe Command Status Code
+ *
+ * @enum xnvme_spec_status_code
+ */
+enum xnvme_spec_status_code {
+	// TODO: Add remaining status codes from spec
```
[the remark this answers is not in the document]
Comment (Baekalfen) [reply include/libxnvme_spec.h:101]: It's not going to be added in this PR. But it's a general issue, that we don't have the status codes and their description in the code.

File: cijoe/tests/bindings/test_cython.py [hunk 6]
```diff
@@ -20,7 +20,30 @@ def test_cython_bindings(cijoe, device, be_opts, cli_args):
...
+def test_cython_bindings_kv(cijoe, device, be_opts, cli_args):
+
+    if be_opts["be"] not in ["linux", "spdk"]:
+        pytest.skip(reason=f"Backend not supported: {be_opts['be']}")
+
+    env = {
+        "XNVME_URI": device["uri"],
+        "XNVME_BE": be_opts["be"],
+        "XNVME_DEV_NSID": f"{device['nsid']}",
+        # "XNVME_HUGETLB_PATH": cijoe.config.options.get("hugetlbfs", {}).get(
```
[the remark this answers is not in the document]
Comment (Baekalfen) [reply cijoe/tests/bindings/test_cython.py:39]: Not in particular

File: lib/xnvme_be.c [hunk 7]
```diff
@@ -369,12 +377,19 @@ xnvme_be_dev_derive_geometry(struct xnvme_dev *dev)
...
+			if (_kvs_geometry(dev)) {
+				XNVME_DEBUG("FAILED: _kvs_geometry");
+				return -EINVAL;
+			}
+			break;
+
 		default:
 			XNVME_DEBUG("FAILED: unhandled csi: 0x%x", dev->ident.csi);
 			return -ENOSYS;
+			break;
```
Comment (safl) [inline lib/xnvme_be.c:390]: Excess break, added to https://github.com/OpenMPDK/xNVMe/issues/288

**Review by safl — APPROVED** [review]
I agree with the suggestions from @karlowich and also found a minor thing. However, to move this along, then I've added an issue to track those things at https://github.com/OpenMPDK/xNVMe/issues/288 and address them after this PR.
Thanks @Baekalfen and @karlowich !
